### PR TITLE
Add test cases for all supported OS

### PR DESCRIPTION
**Pull request by pvogt09**
**By submitting this pull request, I confirm the following:** 

- [x] I have read and understood the [contributors guide](https://github.com/pi-hole/pi-hole/blob/master/CONTRIBUTING.md), as well as this entire template.
- [x] I have made only one major change in my proposed changes.
- [x] I have commented my proposed changes within the code.
- [x] I have tested my proposed changes, and have included unit tests where possible.
- [x] I am willing to help maintain this change if there are issues with it later.
- [x] I give this submission freely and claim no ownership.
- [x] It is compatible with the [EUPL 1.2 license](https://opensource.org/licenses/EUPL-1.1)
- [x] I have squashed any insignificant commits. ([`git rebase`](http://gitready.com/advanced/2009/02/10/squashing-commits-with-rebase.html))

Please make sure you [Sign Off](https://github.com/pi-hole/pi-hole/wiki/How-to-signoff-your-commits.) all commits. Pi-hole enforces the [DCO](https://github.com/pi-hole/pi-hole/wiki/Contributing-to-the-project).

---
**What does this PR aim to accomplish?:**
Add docker containers for all supported OSes to test cases.
Performing the depentent package installation process hangs the test for some unknown reason and is therefore removed again.

**How does this PR accomplish the above?:**
Adds docker containers and updates test setup accordingly.

**What documentation changes (if any) are needed to support this PR?:**
None.

---
* You must follow the template instructions. Failure to do so will result in your pull request being closed.
* Please respect that Pi-hole is developed by volunteers, who can only reply in their spare time.

### CODE REVIEW

**Comment by PromoFaux** [comment]
Thanks. I see the previous test run timed out after 25 mins... I've given it a kick in case it was a glitch

**Comment by PromoFaux** [comment]
Hmm, hanging in the same place 

image

**Comment by pvogt09** [comment]
It seems like something has changed in the tests since March that prevents the tests from running completely.
I assume that it has something to do with the installation of the dependencies in `test_os_check_passes` and `test_os_check_fails`. I have had some trouble with getting that to pass as well back then.
Right now I cannot get any debug output of these fuctions even by adding `set-x` in different places. Do you have any idea where to search for the hang? Maybe it is an input dialog that is not answered?

Something has also changed in the way travis works and I did not take that into account for my fork so I did not see that the tests were not run there, sorry for the inconvenience.

**Comment by pvogt09** [comment]
I have tried to debug the travis run since a hanging build produces not produce any useful output at all with
```
curl -s -X POST -H "Content-Type: application/json" -H "Accept: application/json" -H "Travis-API-Version: 3" -H "Authorization: token ****" -d "{\"quiet\": true}" https://api.travis-ci.com/job/440043639/debug
```
but that gives me only `access denied` (with the correct API token, of course). Do I have to activate debug build anywhere?

Apparently the Travis-not-running problem stems from the new pihole organization global configuration for the builds which prevents travis builds from running on forks because it says
> Skipped as per condition: IF (type = push AND (branch = master OR branch =~ ^release\/v.*$)) OR type IN (pull_request, api, cron)

in the title of the corresponding request for pushes to my fork which stems from https://github.com/pi-hole/.github/blob/master/build-configs/core.yml, maybe that should be changed to allow for builds in forks (@dschaper).

**Comment by dschaper** [comment]
Travis ran 18 days ago with the following errors:

```plain
[gw0] [ 53%] FAILED test/test_automated_install.py::test_FTL_binary_installed_and_responsive_no_errors[ubuntu_18] 

test/test_automated_install.py::test_FTL_binary_installed_and_responsive_no_errors[ubuntu_20] 

[gw0] [ 53%] FAILED test/test_automated_install.py::test_FTL_binary_installed_and_responsive_no_errors[ubuntu_20] 

test/test_automated_install.py::test_IPv6_only_link_local[debian] 

[gw0] [ 54%] PASSED test/test_automated_install.py::test_IPv6_only_link_local[debian] 

test/test_automated_install.py::test_IPv6_only_link_local[debian_9] 

[gw0] [ 54%] PASSED test/test_automated_install.py::test_IPv6_only_link_local[debian_9] 

test/test_automated_install.py::test_IPv6_only_link_local[debian_10] 

[gw0] [ 54%] PASSED test/test_automated_install.py::test_IPv6_only_link_local[debian_10] 

test/test_automated_install.py::test_IPv6_only_link_local[centos7] 

[gw0] [ 55%] PASSED test/test_automated_install.py::test_IPv6_only_link_local[centos7] 

test/test_automated_install.py::test_IPv6_only_link_local[centos] 

[gw0] [ 55%] PASSED test/test_automated_install.py::test_IPv6_only_link_local[centos] 

test/test_automated_install.py::test_IPv6_only_link_local[fedora] 

[gw0] [ 55%] PASSED test/test_automated_install.py::test_IPv6_only_link_local[fedora] 

test/test_automated_install.py::test_IPv6_only_link_local[fedora_31] 

[gw0] [ 56%] PASSED test/test_automated_install.py::test_IPv6_only_link_local[fedora_31] 

test/test_automated_install.py::test_IPv6_only_link_local[fedora_32] 

[gw0] [ 56%] PASSED test/test_automated_install.py::test_IPv6_only_link_local[fedora_32] 

test/test_automated_install.py::test_IPv6_only_link_local[ubuntu_16] 

[gw0] [ 56%] PASSED test/test_automated_install.py::test_IPv6_only_link_local[ubuntu_16] 

[gw1] [ 57%] FAILED test/test_automated_install.py::test_FTL_download_aarch64_no_errors[centos] 

test/test_automated_install.py::test_IPv6_only_link_local[ubuntu_18] 

test/test_automated_install.py::test_FTL_download_aarch64_no_errors[fedora] 

[gw0] [ 57%] PASSED test/test_automated_install.py::test_IPv6_only_link_local[ubuntu_18] 

test/test_automated_install.py::test_IPv6_only_link_local[ubuntu_20] 

[gw0] [ 57%] PASSED test/test_automated_install.py::test_IPv6_only_link_local[ubuntu_20] 

test/test_automated_install.py::test_IPv6_only_ULA[debian] 

[gw0] [ 57%] PASSED test/test_automated_install.py::test_IPv6_only_ULA[debian] 

test/test_automated_install.py::test_IPv6_only_ULA[debian_9] 
```

We don't try to build forks because we can't build forks. Only you can set up travis on your own fork. We can build PRs like this, but you will not have access to our tokens since they are our secret keys.

**Comment by pvogt09** [comment]
I have setup travis in my fork in the "Webhooks" and "Integrations" page and that was working sufficiently well until the organization wide build config was introduced because this now results in the following behaviour:
travis
For this PR, which I am currently investigating in a private branch (`pvogt09:feature/test_all_supported_private`) to not clutter the log and annoy you with notifications too much, every second request to build (marked as red in the picture) comes from a push event to this branch and the build is not run due to the organization config but should run because travis itself is enabled or am I missing something here? The green lines where the build actually ran but failed are triggered manually. Do I have to setup my own organization wide config for travis in my namespace to override that behaviour?

I was trying to debug one of those builds from my fork with my own token (as available on the travis settings page) but that did not succeed. Would I nevertheless need the token from pihole to make this work and if yes how would I be able to figure out why the build process hangs without that?

**Comment by PromoFaux** [comment]
@pvogt09 what OS/environment are you developing on locally? I recently was able to get the tests running locally in a WSL2 instance that I use, though I guess would also work on any linux distro, too.

```
sudo apt install tox
cd /etc/.pihole
tox
```

(the above supposes you also have `docker` installed)

I find it handy to try and run the tests locally, because then I don't need to wait until travis shouts at me to tell me that my tests have failed.

That said, pulling your branch locally and running the tests  - it does pass, but the last 20% or so of tests are quite slow - usually falling over on the `test_os_check_*` tests - like we see on travis. I wonder if I've done a bad job of introducing that test....

**Comment by PromoFaux** [comment]
> I wonder if I've done a bad job of introducing that test....

That all said. Removing those tests - and even removing the code that they were designed to test - the whole process hangs before getting to 100%.

**Comment by pvogt09** [comment]
I am still using Windows 7 with Msys2, so neither WSL nor Docker are available. That is why I run the tests via travis ;-).

The main problem there seems to be updating the package list and installing the dependencies, maybe introducing `travis_wait` as described [here](https://docs.travis-ci.com/user/common-build-problems/#build-times-out-because-no-output-was-received) into the configuration can prevent the timeout as long as this is the only problem. Or reordering the tests in a way that the `test_os_check_*` tests run first before all other tests that do not need that much time might also help.

**Comment by dschaper** [comment]
> I have setup travis in my fork in the "Webhooks" and "Integrations" page and that was working sufficiently well until the organization wide build config was introduced because this now results in the following behaviour:

You will need to change that file in your fork so that it does what you want.

> Would I nevertheless need the token from pihole to make this work

We will not be giving anyone the token from Pi-hole, that's basically our password.

**Comment by dschaper** [comment]
Have you tried making a Pull Request on your fork using your branch?

**Comment by PromoFaux** [comment]
@pvogt09 - might help as a pointer (still trying lots of things) but running locally appeared to show that the unbuntu 20 containers were the culprits 

image

If I remove those from the tox config  then the tests all pass

image

File: test/ubuntu_20.Dockerfile [hunk 1]
```diff
@@ -0,0 +1,16 @@
+FROM buildpack-deps:focal-scm
+
+ENV GITDIR /etc/.pihole
+ENV SCRIPTDIR /opt/pihole
+
+RUN mkdir -p $GITDIR $SCRIPTDIR /etc/pihole
+ADD . $GITDIR
+RUN cp $GITDIR/advanced/Scripts/*.sh $GITDIR/gravity.sh $GITDIR/pihole $GITDIR/automated\ install/*.sh $SCRIPTDIR/
+ENV PATH /usr/local/sbin:/usr/local/bin:/usr/sbin:/usr/bin:/sbin:/bin:$SCRIPTDIR
```
Comment (PromoFaux) [inline test/ubuntu_20.Dockerfile:9]: @pvogt09 this is the solution. With a lot of help from @dschaper  (thanks!) we managed to track it down to a whiptail being displayed - and waiting for user input - when the install test required that `install_dependent_packages` was called for `dnsutils`

Adding the below sorts it:


```suggestion
ENV PATH /usr/local/sbin:/usr/local/bin:/usr/sbin:/usr/bin:/sbin:/bin:$SCRIPTDIR
ENV DEBIAN_FRONTEND=noninteractive
```

**Comment by PromoFaux** [comment]
@pvogt09 , I assumed that my commit here would request a PR build - but it does not seem to have done. You might want to add it yourself and force push

**Comment by pvogt09** [comment]
Great, thanks for the support.

Should I just force push your commit to trigger another build?

It is strange that this works when set in the dockerfile, because I have added `export DEBIAN_FRONTEND=noninteractive` to the `Pihole.run` call during my tests as well and that did not solve anything.

Playing around with `travis_wait` obviously did not lead to anything useful.

> 
> 
> Have you tried making a Pull Request on your fork using your branch?

Making a PR to my own fork starts the travis build sucessfully.

**Comment by PromoFaux** [comment]
> Should I just force push your commit to trigger another build?

You could try. If that does not work, just reset back to the commit before mine, make the change yourself on a fresh commit, and then force push

**Review by PromoFaux — APPROVED** [review]
Happy to merge this one as is - after spending the afternoon picking it apart.

**Pull request merged**